src/Lean/Meta/Tactic/Try/Collect.lean

```diff
@@ -0,0 +1,210 @@
+/-
+Copyright (c) 2025 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Leonardo de Moura
+-/
+prelude
+import Init.Try
+import Lean.Meta.Tactic.LibrarySearch
+import Lean.Meta.Tactic.Util
+import Lean.Meta.Tactic.Grind.Cases
+import Lean.Meta.Tactic.Grind.EMatchTheorem
+
+namespace Lean.Meta.Try.Collector
+
+structure InductionCandidate where
+  fvarId : FVarId
+  val    : InductiveVal
+
+structure FunIndCandidate where
+  funIndDeclName : Name
+  majors : Array FVarId
+  deriving Hashable, BEq
+
+structure Result where
+  /-- All constant symbols occurring in the gal. -/
+  allConsts : Std.HashSet Name  := {}
+  /-- Unfolding candiates. -/
+  unfoldCandidates : Std.HashSet Name  := {}
+  /-- Equation function candiates. -/
+  eqnCandidates : Std.HashSet Name  := {}
+  /-- Function induction candidates. -/
+  funIndCandidates : Std.HashSet FunIndCandidate := {}
+  /-- Induction candidates. -/
+  indCandidates : Array InductionCandidate := #[]
+  /-- Relevant declarations by `libSearch` -/
+  libSearchResults : Std.HashSet (Name × Grind.EMatchTheoremKind) := {}
+
+structure Context where
+  config : Try.Config
+
+abbrev M := ReaderT Context <| StateRefT Result MetaM
+
+def getConfig : M Try.Config := do
+  return (← read).config
+
+def saveConst (declName : Name) : M Unit := do
+  modify fun s => { s with allConsts := s.allConsts.insert declName }
+
+/-- Returns `true` if `declName` is in the module being compiled. -/
+def inCurrentModule (declName : Name) : CoreM Bool := do
+  return ((← getEnv).getModuleIdxFor? declName).isNone
+
+def getFunInductName (declName : Name) : Name :=
+  declName ++ `induct
+
+def getFunInduct? (declName : Name) : MetaM (Option Name) := do
+  let .defnInfo _ ← getConstInfo declName | return none
+  try
+    let result ← realizeGlobalConstNoOverloadCore (getFunInductName declName)
```

Comment:
We really ought to cache realization failures soon! Imagine a theorem fails to realize (and slowly so), and it appears a few times in the goal. Then `try` becomes very slow without any feedback. A simple non-persistent extension should work. 

Ah, but last time we talked about the the conclusion was to wait for parallelism to deal with lazy generation and see if it's still an issue after that